curp/src/channel/key_spmc.rs

```diff
@@ -0,0 +1,292 @@
+    fn new(keys: Vec<K>, message: M) -> Self {
+        Self {
+            inner: Arc::new(KeysMessageInner::new(keys, message)),
+        }
+    }
+
+    /// modify data in the message if necessary
+    pub(crate) fn keys(&self) -> &[K] {
+        self.inner.keys()
+    }
+
+    /// map the inner message to a closure
+    pub(crate) fn map_msg<R, F: FnOnce(&mut M) -> R>(&self, f: F) -> R {
+        self.inner.map_msg(f)
+    }
+}
+
+impl<K: Eq + Hash + Clone + ConflictCheck, M> KeybasedChannel<SpmcKeysMessage<K, M>> {
+    /// Insert successors and proceeders info and return if it conflicts with any previous message
+    fn insert_graph(&mut self, new_km: SpmcKeysMessage<K, M>) -> bool {
+        let proceeder_cnt: u64 = self
+            .successor
+            .iter_mut()
+            .filter_map(|(k, v)| {
+                super::keys_conflict(k.keys(), new_km.keys()).then(|| {
+                    let _ignore = v.insert(new_km.clone());
+                    1
+                })
+            })
+            .sum();
+        // the message can only be inserted once, so we ignore the return value
+        let _ignore = self.successor.insert(new_km.clone(), HashSet::new());
+
+        if proceeder_cnt == 0 {
+            false
+        } else {
+            // the message can only be inserted once, so we ignore the return value
+            let _ignore2 = self.proceeder.insert(new_km, proceeder_cnt);
+            true
+        }
+    }
+
+    /// Append a key and message to this `KeybasedChannel`
+    fn append(&mut self, keys: &[K], msg: M) {
+        let km = SpmcKeysMessage::new(keys.to_vec(), msg);
+        if !self.insert_graph(km.clone()) {
+            self.inner.push_back(km);
+            self.new_msg_event.notify(1);
+        }
+    }
+
+    /// Move a message for the `key` from pending to inner
+    fn mark_done(&mut self, km: &SpmcKeysMessage<K, M>) {
+        let ready_cnt = if let Some(successor) = self.successor.remove(km) {
+            successor
+                .into_iter()
+                .map(|s| {
+                    let (no_proceeder, has_pre) = if let Some(s_p) = self.proceeder.get_mut(&s) {
+                        *s_p = s_p.overflow_sub(1);
+                        (*s_p == 0, true)
+                    } else {
+                        (true, false)
+                    };
+
+                    if no_proceeder && has_pre {
+                        let _ignore_removed = self.proceeder.remove(&s);
+                    }
+
+                    if no_proceeder {
+                        self.inner.push_back(s);
+                        1
+                    } else {
+                        0
+                    }
+                })
+                .sum()
+        } else {
+            0
+        };
+
+        if ready_cnt > 0 {
+            self.new_msg_event.notify(ready_cnt);
+        }
+    }
+}
+
+/// The Sender for the `KeybasedChannel`
+pub(crate) struct SpmcKeybasedSender<K, M> {
```

Comment:
!impl Sync

Reply:
Any type that not implements `Sync` is not syncable.